### PR TITLE
Match urls with query params event if the params are not in the same order

### PR DESCRIPTION
**Pull request by shmilky**
See https://github.com/wheresrhys/fetch-mock/issues/227

* For cases where the url is a valid one we'll do a less basic equality check by separating the query params and compare them separately

* Added basic test

* Add query-string package to parse the query strings.

### CODE REVIEW

File: src/compile-route.js [hunk 1]
```diff
@@ -62,6 +63,55 @@ function areHeadersEqual (currentHeader, expectedHeader) {
     return true;
 }
 
+const pathPattern = /^(https?:\/\/)?([\da-z\.-]+)\.([a-z\.]{2,6})([\/\w \.-]*)*\/?$/;
+const queryPattern = /^(\?)?([0-9A-Za-z]+=[0-9A-Za-z]+(\&)?)+$/;
+
+function getUrlSections (url) {
```
Comment (wheresrhys) [inline src/compile-route.js]: can't you use `URL.parse()` here https://nodejs.org/api/url.html

File: src/compile-route.js [hunk 2]
```diff
@@ -62,6 +63,55 @@ function areHeadersEqual (currentHeader, expectedHeader) {
...
+    const pathPart = splitUrl[0];
+    const queryPart = splitUrl.length > 1 ? splitUrl.slice(1).join('?') : '';
+
+    return {
+        path: pathPart,
+        query: queryPart
+    }
+}
+
+function isValidUrl (checkedUrl) {
```
Comment (wheresrhys) [inline src/compile-route.js]: Similar to above, if you use the URL class, there will probably be no need for this custom validation

File: src/compile-route.js (hunk 1 above)
Comment (vitalets) [reply src/compile-route.js]: Maybe consider url normalization package? e.g. https://github.com/sindresorhus/normalize-url

**Comment by wheresrhys** [comment]
I'm writing v6 of the library. I intend to support query string reordering via a similar mechanism to specifying headers i.e. if query order is unimportant, they should not be added in the string that matches the url, but rather passed in as a map in the options. This feels less magic and it's far clearer how this behaviour would blend with the various different methods of matching a url e.g. the basic url matcher could be a regex, but still support unordered query string matching.

Thanks for raising this and for the initial PR though

**Pull request closed without merging**